File: getting_started/files/sequential.R
# assumed to be sequential.R
library(foreach)


# this function approximates pi by throwing random points into a square
# it is used here to demonstrate a function that takes a bit of time
approximate_pi <- function() {
  # number of points to use
  n <- 2000000

  # generate n random points in the square
  x <- runif(n, -1.0, 1.0)
  y <- runif(n, -1.0, 1.0)

  # count the number of points that are inside the circle
  n_in <- sum(x^2 + y^2 < 1.0)

  4 * n_in / n
}


foreach (i=1:100, .combine=c) %do% {
  approximate_pi()
}

File: getting_started/files/sequential.c
// assumed to be sequential.c
#include <stdio.h>
#include <time.h>

double approximate_pi(void) {
    const int n = 2000000;
    int n_in = 0;
    for (int i = 0; i < n; i++) {
        double x = (double)rand() / RAND_MAX * 2.0 - 1.0;
        double y = (double)rand() / RAND_MAX * 2.0 - 1.0;
        if (x * x + y * y < 1.0)
            n_in++;
    }
    return 4.0 * n_in / n;
}

int main(void) {
    srand(time(NULL));
    for (int i = 0; i < 100; i++)
        approximate_pi();
    return 0;
}
